<a href="https://colab.research.google.com/github/hyrinn/Deep_Learning/blob/main/4%EC%A3%BC%EC%B0%A8_%EC%98%B5%ED%8B%B0%EB%A7%88%EC%9D%B4%EC%A0%80_%EC%84%B1%EB%8A%A5_%EB%B9%84%EA%B5%90.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**[4주차]과제**
- 아래의 과제1), 과제2)의 코드를 완성하시오.
- 모든 코드의 결과를 출력하여, .ipynb의 링크를 **[4주차]/[4주차]과제**에 제출하시오.\
(실습 제출 예시: 4주차_2020XXXX_이름.ipynb 코드 링크)

In [ ]:
# 학번, 이름 출력
print(f'학번 : 2243186 이름 : 조효린')

학번 : 2243186 이름 : 조효린


In [ ]:
# google drive 연결
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# 구글 드라이브에 데이터를 넣고 path 연결 필요
data_path = '/content/drive/MyDrive/deep/week_4'

## 과제 1
- 케라스를 이용하여 은닉층이 2개인 MLP를 생성하고 옵티마이저를 Adam, RMSprop, Adadelta, Adagrad로 변경해보면서 성능을 측정한 뒤 어떤 방법이 가장 좋은 성능을 보이는지 확인하고, 각 방법에 대해서 서술하시오.
- 은닉층의 노드 수는 512, 256로 설정
- 은닉층의 Activation Function은 ReLU로 설정
- 모델 레이어 추가  
model.add(tf.keras.layers.Dense(노드갯수, activation='activation_func', kernel_initializer=initializer))  
<출력 예시>  
테스트 정확도 : 0.2779000282287598


In [ ]:
# 런타임 재시작 필요
!pip install scikeras==0.13.0 scikit-learn==1.4.2

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.2/12.2 MB 54.8 MB/s eta 0:00:00
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.6.1
    Uninstalling scikit-learn-1.6.1:
      Successfully uninstalled scikit-learn-1.6.1
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
umap-learn 0.5.11 requires scikit-learn>=1.6, but you have scikit-learn 1.4.2 which is incompatible.
hdbscan 0.8.41 requires scikit-learn>=1.6, but you have scikit-learn 1.4.2 which is incompatible.


In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
from sklearn.model_selection import GridSearchCV
from scikeras.wrappers import KerasClassifier, KerasRegressor
import os
import numpy as np

In [ ]:
# 실험 결과 재현을 위한 seed 고정
SEED = 4
os.environ['PYTHONHASHSEED'] = str(SEED)
os.environ['TF_DETERMINISTIC_OPS'] = '1'
tf.random.set_seed(SEED)
np.random.seed(SEED)
tf.random.set_seed(SEED)
initializer = tf.keras.initializers.GlorotUniform(seed=SEED)

In [ ]:
# 데이터세트 준비
(train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.mnist.load_data()

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 3s 0us/step


In [ ]:
# 데이터 전처리
train_images = train_images.reshape((60000, 784))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 784))
test_images = test_images.astype('float32') / 255

train_labels = tf.keras.utils.to_categorical(train_labels)
test_labels = tf.keras.utils.to_categorical(test_labels)

In [ ]:
# 신경망 모델 구축
def build_model():
    network = tf.keras.models.Sequential()

    # 은닉층 생성
    # 여기에 코드를 작성해주세요

    network.add(tf.keras.layers.Dense(512, activation='relu', input_shape=(784,),kernel_initializer=initializer))
    network.add(tf.keras.layers.Dense(10, activation='sigmoid', kernel_initializer=initializer))

    # 옵티마이저 설정 필요
    network.compile(optimizer='adam',
                   loss='mse',
                   metrics=['accuracy'])
    return network

In [ ]:
# model 생성
model = build_model()

In [ ]:
result = model.fit(train_images, train_labels, epochs=5, batch_size=64)

Epoch 1/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 10s 10ms/step - accuracy: 0.9129 - loss: 0.0155
Epoch 2/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 10s 11ms/step - accuracy: 0.9675 - loss: 0.0060
Epoch 3/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 11s 12ms/step - accuracy: 0.9790 - loss: 0.0041
Epoch 4/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 9s 10ms/step - accuracy: 0.9845 - loss: 0.0031
Epoch 5/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 11s 11ms/step - accuracy: 0.9880 - loss: 0.0024


In [ ]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9800 - loss: 0.0038


In [ ]:
print('테스트 정확도 : ', test_acc)

테스트 정확도 :  0.9800000190734863


### 각 옵티마이저의 성능 작성
- RMSprop : 0.9722999930381775
- Adagrad : 0.555400013923645
- Adadelta : 0.2955999970436096
- Adam : 0.9800000190734863

### 가장 좋은 성능의 옵티마이저 작성
- Adam

### 옵티마이저 서술
- RMSprop : 학습 중 기울기의 제곱 이동 평균을 사용하여 각 매개변수의 학습률을 적응적으로 조정하는 최적화 알고리즘
- Adagrad :  학습률을 파라미터별로 다르게 조정(Adaptive)하여 적용하는 경사하강법 기반의 옵티마이저
- Adadelta : AdaGrad의 학습률 감소 문제를 해결하기 위해 제안된, 경사하강법 기반의 어댑티브 학습률 최적화 알고리즘
- Adam : 딥러닝 모델의 학습 과정에서 손실 함수의 최솟값을 효율적으로 찾기 위해 가중치를 업데이트하는 최적화 알고리즘

## 과제 2 MLP 모델의 정확도(Accuracy)를 71이상 달성하시오.  
- 해당 diabetes.csv 데이터는 Pima 인디언의 의료 기록과 각 환자가 5년 이내에 당뇨병 발병 여부가 저장되어 있다.
- 현재 모델은 1 : 당뇨병 양성, 0 : 당뇨병 음성을 분류하는 MLP 모델이다.
- 해당 모델의 하이퍼 파라미터를 수정하여 정확도 71 이상을 달성하여라.
- 하이퍼 파라미터 : Activation Function, optimizer, epochs, batch size 등


In [ ]:
!pip install pandas

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adagrad, Adadelta, RMSprop, Adam
import pandas as pd
import numpy as np
import os

In [ ]:
# 데이터 load
dataset = pd.read_csv(os.path.join(data_path, '/content/drive/MyDrive/diabetes.csv'))

X = dataset.iloc[:, 0:8].values
y = dataset.iloc[:, 8].values

In [ ]:
# 수정 x
def configure_tf_determinism():
    # TensorFlow의 병렬성 감소
    tf.config.experimental.enable_op_determinism()

In [ ]:
# 실험 결과 재현을 위한 seed 고정
SEED = 4
os.environ['PYTHONHASHSEED'] = str(SEED)
os.environ['TF_DETERMINISTIC_OPS'] = '1'
tf.random.set_seed(SEED)
np.random.seed(SEED)
tf.random.set_seed(SEED)
initializer = tf.keras.initializers.GlorotUniform(seed=SEED)

In [ ]:
# 모델 생성
def build_model():
    model = Sequential()
    # 수정 가능
    model.add(tf.keras.layers.Dense(12, input_dim=8, activation='sigmoid', kernel_initializer=initializer))
    model.add(tf.keras.layers.Dense(8, activation='sigmoid', kernel_initializer=initializer))

    # 수정 x
    model.add(tf.keras.layers.Dense(1, activation='sigmoid', kernel_initializer=initializer))

    model.compile(loss='binary_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

    return model

In [ ]:
# 수정 x
configure_tf_determinism()

In [ ]:
model = build_model()

In [ ]:
model.fit(X, y, epochs=30, batch_size=40)

Epoch 1/30
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7005 - loss: 0.5648 
Epoch 2/30
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7005 - loss: 0.5644 
Epoch 3/30
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7005 - loss: 0.5643 
Epoch 4/30
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7005 - loss: 0.5642 
Epoch 5/30
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7005 - loss: 0.5640 
Epoch 6/30
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6992 - loss: 0.5638 
Epoch 7/30
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7005 - loss: 0.5637 
Epoch 8/30
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7005 - loss: 0.5635 
Epoch 9/30
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7005 - loss: 0.5633
Epoch 10/30
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.7005 - loss: 0.5631
Epoch 11/30
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7005 - loss: 0.5630
Epoch 12/30
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7

In [ ]:
_, accuracy = model.evaluate(X, y)

24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7214 - loss: 0.5585 


In [ ]:
print('정확도: %.2f' % (accuracy*100))

정확도: 72.14
